File: packages/mcp/src/validation.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Shared §2.6.1 submission validation.
 * Used by both @atrib/log-dev and log-node servers.
 *
 * Steps 2–5 of §2.6.1 plus required-field presence checks.
 * Step 1 (Ed25519 signature verification) is intentionally excluded here.
 * it lives in @atrib/verify and would create a circular workspace dependency.
 * Step 6 (idempotency) is state-dependent and handled by each server's
 * proof-cache / storage layer.
 *
 * The 10-minute future-skew window (Step 4) matches §2.6.1's log-server
 * tolerance. Client-side record verification (§1.4.3) uses a tighter
 * 5-minute window; see packages/mcp/src/signing.ts for that path.
 */

import type { AtribRecord } from './types.js'
import { isValidEventTypeUri } from './types.js'
const SPEC_VERSION = 'atrib/1.0'
const MAX_FUTURE_SKEW_MS = 10 * 60 * 1000 // spec 2.6.1 Step 4: 10 minutes

export interface ValidationResult {
  ok: boolean
  status?: number
  error?: string
}

/**
 * Validate a §2.6.1 submission body (Steps 2–5 + required fields).
 * Returns `{ ok: true }` if all checks pass, or `{ ok: false, status, error }`
 * if any check fails.
 */
export function validateSubmission(record: Partial<AtribRecord>): ValidationResult {
  // Guard: reject non-object input (null, undefined, primitives)
  if (!record || typeof record !== 'object') {
    return { ok: false, status: 400, error: `spec_version must be '${SPEC_VERSION}'` }
  }

  // Step 2: spec_version must be 'atrib/1.0'
  if (record.spec_version !== SPEC_VERSION) {
    return { ok: false, status: 400, error: `spec_version must be '${SPEC_VERSION}'` }
  }

  // Step 3: event_type must be a syntactically-valid absolute URI per spec 1.4.5.
  // Recognition (whether the URI is in atrib normative set) is informational only.
  if (!isValidEventTypeUri(record.event_type)) {
    return {
      ok: false,
      status: 400,
      error: 'event_type must be a syntactically-valid absolute URI',
    }
  }

  // Step 4: timestamp must be present and not more than 10 minutes in the future
  if (typeof record.timestamp !== 'number' || !Number.isInteger(record.timestamp) || record.timestamp < 0) {
    return { ok: false, status: 400, error: 'timestamp must be a non-negative integer' }
  }
  if (record.timestamp - Date.now() > MAX_FUTURE_SKEW_MS) {
    return { ok: false, status: 400, error: 'timestamp is more than 10 minutes in the future' }
  }

  // Step 5: context_id must be exactly 32 lowercase hex chars
  if (typeof record.context_id !== 'string' || !/^[0-9a-f]{32}$/.test(record.context_id)) {
    return { ok: false, status: 400, error: 'context_id must be 32 lowercase hex characters' }
  }

  // Required string fields for record_hash computation
  for (const field of ['creator_key', 'chain_root', 'content_id', 'signature'] as const) {
    if (typeof record[field] !== 'string') {
      return { ok: false, status: 400, error: `${field} is required and must be a string` }
    }
  }

  // T13 (Phase 5): chain_root must match the spec §1.2.3 format —
  // either the genesis form 'sha256:' + 64-hex of SHA-256(context_id),
  // or a sha256: prefix + 64-hex of a prior record_hash. The format
  // check rejects malformed inputs (empty string, missing prefix,
  // wrong digest length) at the API boundary so they can't pollute
  // the log.
  if (!/^sha256:[0-9a-f]{64}$/.test(record.chain_root as string)) {
    return { ok: false, status: 400, error: 'chain_root must match sha256:<64-hex>' }
  }

  // content_id follows the same shape per §1.2.5.
  if (!/^sha256:[0-9a-f]{64}$/.test(record.content_id as string)) {
    return { ok: false, status: 400, error: 'content_id must match sha256:<64-hex>' }
  }

  // session_token is optional and OMITTED when absent (§1.3), but if present must be a string
  if ('session_token' in record && typeof record.session_token !== 'string') {
    return { ok: false, status: 400, error: 'session_token must be a string when present' }
  }

  // informed_by is optional (D041 / §1.2.7). When present: array of
  // `sha256:<64-hex>` strings, each pointing at a prior record_hash.
  // Order matters for canonicalization but is preserved as-given (verifiers
  // tolerate any order; the JCS form embeds the array verbatim).
  if ('informed_by' in record) {
    const ib = record.informed_by
    if (!Array.isArray(ib)) {
      return { ok: false, status: 400, error: 'informed_by must be an array when present' }
    }
    for (const ref of ib) {
      if (typeof ref !== 'string' || !/^sha256:[0-9a-f]{64}$/.test(ref)) {
        return { ok: false, status: 400, error: 'informed_by entries must each match sha256:<64-hex>' }
      }
    }
  }

  return { ok: true }
}

File: python/src/atrib/validation.py
# SPDX-License-Identifier: Apache-2.0
"""§2.6.1 submission validation (client-side parity).

Port of ``packages/mcp/src/validation.ts``: Steps 2–5 plus required-field
presence checks. Step 1 (Ed25519 verification) lives in
:func:`atrib.signing.verify_record`; Step 6 (idempotency) is server state.
The 10-minute future-skew window matches the log server's tolerance
(client-side §1.4.3 verification uses the tighter 5-minute window).
"""

from __future__ import annotations

import re
import time
from collections.abc import Mapping
from dataclasses import dataclass

from .types import SPEC_VERSION, is_valid_event_type_uri

_MAX_FUTURE_SKEW_MS = 10 * 60 * 1000
_SHA256_REF_RE = re.compile(r"^sha256:[0-9a-f]{64}\Z")
_CONTEXT_ID_RE = re.compile(r"^[0-9a-f]{32}\Z")


@dataclass(frozen=True)
class ValidationResult:
    ok: bool
    status: int | None = None
    error: str | None = None


def _reject(error: str) -> ValidationResult:
    return ValidationResult(ok=False, status=400, error=error)


def _is_integral_number(value: object) -> bool:
    """Port of JS Number.isInteger over parsed-JSON values: ints and
    integral floats qualify; bools (Python ints) do not — JS `typeof true`
    is not 'number'."""
    if isinstance(value, bool):
        return False
    if isinstance(value, int):
        return True
    return isinstance(value, float) and value.is_integer()


def validate_submission(
    record: object, *, now_ms: int | None = None
) -> ValidationResult:
    """Validate a §2.6.1 submission body. ``now_ms`` injects the clock so
    corpus harnesses can pin the manifest's reference time."""
    if not isinstance(record, Mapping):
        return _reject(f"spec_version must be '{SPEC_VERSION}'")

    # Step 2: spec_version.
    if record.get("spec_version") != SPEC_VERSION:
        return _reject(f"spec_version must be '{SPEC_VERSION}'")

    # Step 3: event_type syntactic URI validity (extensions pass).
    if not is_valid_event_type_uri(record.get("event_type")):
        return _reject("event_type must be a syntactically-valid absolute URI")

    # Step 4: timestamp integer, not >10min in the future. Integral floats
    # pass: JSON "1.0e12" parses to the same JS number as "1000000000000",
    # and the TS reference (Number.isInteger) accepts it — Python must not
    # be stricter on identical wire bytes.
    timestamp = record.get("timestamp")
    if not _is_integral_number(timestamp) or not isinstance(timestamp, (int, float)):
        return _reject("timestamp must be a non-negative integer")
    if timestamp < 0:
        return _reject("timestamp must be a non-negative integer")
    now = int(time.time() * 1000) if now_ms is None else now_ms
    if timestamp - now > _MAX_FUTURE_SKEW_MS:
        return _reject("timestamp is more than 10 minutes in the future")

    # Step 5: context_id exactly 32 lowercase hex chars.
    context_id = record.get("context_id")
    if not isinstance(context_id, str) or not _CONTEXT_ID_RE.match(context_id):
        return _reject("context_id must be 32 lowercase hex characters")

    # Required string fields for record_hash computation.
    for field in ("creator_key", "chain_root", "content_id", "signature"):
        if not isinstance(record.get(field), str):
            return _reject(f"{field} is required and must be a string")

    if not _SHA256_REF_RE.match(str(record.get("chain_root"))):
        return _reject("chain_root must match sha256:<64-hex>")

    if not _SHA256_REF_RE.match(str(record.get("content_id"))):
        return _reject("content_id must match sha256:<64-hex>")

    if "session_token" in record and not isinstance(record.get("session_token"), str):
        return _reject("session_token must be a string when present")

    if "informed_by" in record:
        informed_by = record.get("informed_by")
        if not isinstance(informed_by, list):
            return _reject("informed_by must be an array when present")
        for ref in informed_by:
            if not isinstance(ref, str) or not _SHA256_REF_RE.match(ref):
                return _reject("informed_by entries must each match sha256:<64-hex>")

    return ValidationResult(ok=True)
